IGNITE-13743 Add defragmentation JMX bean

SammyVimes opened the pull request.

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/CachePartitionDefragmentationManager.java:
```diff
@@ -1013,57 +1026,86 @@ public void createNewCacheDataStore(GridSpinBusyLock busyLock) {
+        }
+
+        /** */
+        public synchronized void onStart(Set<CacheGroupContext> scheduledGroups, int partitions) {
```
If everything is private now then this method should be private as well

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/CachePartitionDefragmentationManager.java:
```diff
@@ -1072,61 +1114,76 @@ public synchronized void onCacheGroupFinish(CacheGroupContext grpCtx) {
-            if (!finishedGroups.isEmpty()) {
-                sb.append("Defragmentation is completed for cache groups:\n");
+        /** */
+        private synchronized Status copy() {
```
So, who calls it? All I see is this:

public Status status() {
    return status;
}

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/CachePartitionDefragmentationManager.java:
```diff
@@ -1072,61 +1114,76 @@ public synchronized void onCacheGroupFinish(CacheGroupContext grpCtx) {
-                for (Map.Entry<CacheGroupContext, DefragmentationCacheGroupProgress> entry : finishedGroups.entrySet()) {
-                    sb.append("    ").append(entry.getKey().cacheOrGroupName()).append(" - ");
+        /** */
+        public long getStartTs() {
```
Auto-generated names? Can you please change them? We don't use abbreviations in method names.

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/CachePartitionDefragmentationManager.java:
```diff
@@ -1072,61 +1114,76 @@ public synchronized void onCacheGroupFinish(CacheGroupContext grpCtx) {
+                finishTs,
+                new HashSet<>(scheduledGroups),
+                new HashMap<>(progressGroups),
+                new HashMap<>(finishedGroups),
```
Underlying data structures are not thread-safe (and they are mutable). You have to make a deep copy.

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/DefragmentationMXBeanImpl.java:
```diff
@@ -0,0 +1,85 @@
+    /** Defragmentation manager. */
+    private final IgniteDefragmentation defragmentation;
+
+    public DefragmentationMXBeanImpl(GridKernalContext ctx) {
```
Comment is missing.

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/IgniteDefragmentation.java:
```diff
@@ -0,0 +1,341 @@
+        private final long totalElapsedTime;
+
+        public DefragmentationStatus(
+                Map<String, CompletedDefragmentationInfo> completedCaches,
```
Too many spaces, we use 4, not 8 =)

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/IgniteDefragmentationImpl.java:
```diff
@@ -0,0 +1,223 @@
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    public IgniteDefragmentationImpl(GridKernalContext ctx) {
```
I see no comment here as well.

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/IgniteDefragmentationImpl.java:
```diff
@@ -0,0 +1,223 @@
+    @Override public boolean inProgress() {
+        final Status status = getStatus();
+
+        return status != null && status.getFinishTs() == 0;
```
So, if status is null then inProgress is false.
1) is that correct?
2) how is it possible?

Is this the case when we're not in maintanance mode or there's no defragmentation scheduled?

SammyVimes replied:
Yes, status is null if the node is not in the maintenance mode/no defragmentation scheduled

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/visor/defragmentation/VisorDefragmentationTask.java:
```diff
@@ -120,91 +110,71 @@ protected VisorDefragmentationJob(@Nullable VisorDefragmentationTaskArg arg, boo
+                case SUCCESS_SUPERSEDED_PREVIOUS:
+                    message = "Scheduling completed successfully. Previously scheduled task has been removed.";
+                    break;
+                case SUCCESS:
```
Not enough empty lines. Applies for all changes

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/visor/defragmentation/VisorDefragmentationTask.java:
```diff
@@ -120,91 +110,71 @@ protected VisorDefragmentationJob(@Nullable VisorDefragmentationTaskArg arg, boo
-            return new VisorDefragmentationTaskResult(true, defrgMgr.status());
+            try {
+                return new VisorDefragmentationTaskResult(true, defragmentation.status().toString());
+            } catch (IgniteCheckedException e) {
```
Formatting is wrong

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/visor/defragmentation/VisorDefragmentationTask.java:
```diff
@@ -120,91 +110,71 @@ protected VisorDefragmentationJob(@Nullable VisorDefragmentationTaskArg arg, boo
+            try {
+                final IgniteDefragmentation.CancelResult cancelResult = defragmentation.cancel();
+
+                String message;
```
Please rename to msg.

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/visor/defragmentation/VisorDefragmentationTask.java:
```diff
@@ -120,91 +110,71 @@ protected VisorDefragmentationJob(@Nullable VisorDefragmentationTaskArg arg, boo
-                    return new VisorDefragmentationTaskResult(false, "Exception occurred: " + e.getMessage());
-                }
+                return new VisorDefragmentationTaskResult(true, message);
+            } catch (IgniteCheckedException e) {
```
Wrong formatting.

ibessonov commented on modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/DefragmentationMXBeanTest.java:
```diff
@@ -0,0 +1,277 @@
+                startGrid(0, cfgOp);
+            }
+            catch (Exception e) {
+                // No-op.
```
Something's wrong here.

ibessonov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/CachePartitionDefragmentationManager.java:
```diff
@@ -232,7 +231,30 @@ public void beforeDefragmentation() throws IgniteCheckedException {
+            oldStores.put(grpId, oldCacheDataStores);
+        }
+
+        int partitionCount = oldStores.values().stream().mapToInt(List::size).sum();
```
So, what's up with index files?

ibessonov commented on modules/core/src/main/java/org/apache/ignite/mxbean/DefragmentationMXBean.java:
```diff
@@ -0,0 +1,73 @@
+     * @return {@code true} if defragmentation is scheduled, {@code false} otherwise.
+     */
+    @MXBeanDescription("Schedule defragmentation.")
+    public boolean schedule(@MXBeanParameter(name = "cacheNames", description = "Names of caches to run defragmentation on.") String cacheNames);
```
Please add "comma separated" into description

ibessonov committed:
IGNITE-13814 restorePartitionStates moved to sys pool instead of striped pool. - Fixes #8542.

Signed-off-by: Sergey Chugunov <sergey.chugunov@gmail.com>

isapego committed:
IGNITE-13801: Fix Ab Initio related ODBC issues

This closes #8528

mrk-andreev committed:
IGNITE-13713 Add target encoding preprocessor (#8466)

mrk-andreev committed:
IGNITE-13714 Add catboost inference integration (#8489)

sergeyuttsel committed:
IGNITE-13353 Got rid of unnecessary rebalance on starting new cache.

Signed-off-by: Slava Koptilin <slava.koptilin@gmail.com>

ibessonov committed:
IGNITE-13823 WAL iterator WRITE permission requirement removed. - Fixes #8549.

Signed-off-by: Sergey Chugunov <sergey.chugunov@gmail.com>

sk0x50 committed:
IGNITE-13450 [MINOR] Added missed javadoc for EVT_CACHE_QUERY_EXECUTED event.

ibessonov committed:
IGNITE-13786 Add defragmentation-specific B+Tree optimizations - Fixes #8560.

Signed-off-by: Alexey Goncharuk <alexey.goncharuk@gmail.com>

ptupitsyn committed:
IGNITE-13826 .NET: Add RendezvousAffinityFunction.BackupFilter

Add RendezvousAffinityFunction.BackupFilter with a single predefined implementation that delegates to Java: ClusterNodeAttributeAffinityBackupFilter.

ibessonov committed:
IGNITE-13833 More versions added to PersistenceBasicCompatibilityTest - Fixes #8562.

Signed-off-by: Ilya Kasnacheev <ilya.kasnacheev@gmail.com>

ibessonov committed:
IGNITE-13832 Proper handling of interrupted exceptions in disco-notifier-worker. - Fixes #8561.

Signed-off-by: Sergey Chugunov <sergey.chugunov@gmail.com>

ibessonov committed:
IGNITE-13101 Metastore should complete all write futures during stop and prohibit creating new ones - Fixes #8554.

Signed-off-by: Sergey Chugunov <sergey.chugunov@gmail.com>

tkalkirill committed:
IGNITE-13815 Remove ability to delete segments from the middle of WAL archive - Fixes #8545.

Signed-off-by: Ilya Kasnacheev <ilya.kasnacheev@gmail.com>

SammyVimes committed:
IGNITE-12892 WAL archive size configuration made more clear - Fixes #8550.

Signed-off-by: Sergey Chugunov <sergey.chugunov@gmail.com>

zstan committed:
IGNITE-13838 IgniteSqlSplitterSelfTest fixes various tests - Fixes #8565.

Signed-off-by: Ilya Kasnacheev <ilya.kasnacheev@gmail.com>

Nikita-tech-writer committed:
ignite docs: fixing a broken documentation link

SammyVimes committed:
IGNITE-13743 Defragmentation JMX API

SammyVimes committed:
IGNITE-13743 Fix javadoc

SammyVimes committed:
IGNITE-13743 Fix javadoc
